### PR TITLE
When available, use BigInt to format large numbers.

### LINKED ISSUE #24: The decimal format (d) isn’t observed if x ≥ 1e21.
For example:

``` js
var f = d3.format(",d");
f(1e20); // "100,000,000,000,000,000,000"
f(1e21); // "1e+21"
```

Because:

``` js
(1e20).toString(10); // "100000000000000000000"
(1e21).toString(10); // "1e+21"
```

Related #21.

### PR DESCRIPTION
**Pull request by mbostock**
Supersedes #97. Fixes #24.

We could, if desired, use *number*.toPrecision for values 1e100… but I think with adoption for BigInt increasing, that’s probably not necessary. And really, if you’re going to be formatting large integers, you probably want to use BigInt’s toLocaleString directly for precision.

### CODE REVIEW

**Comment by mbostock** [comment]
(BigInt is part of ES2020, which required an ESLint upgrade… so I updated the other dependencies at the same time. But this is still backwards-compatible and will behave nicely if BigInt is unavailable.)

**Comment by mbostock** [comment]
This can be released as a patch version.

**Comment by mbostock** [comment]
This could be released as a patch version… but I think it’s probably safer to bundle this together with the other “breaking” change to the default minus sign #103.

File: test/format-type-d-test.js [hunk 1]
```diff
@@ -77,7 +77,9 @@ tape("format(\",d\") can group thousands", function(test) {
   test.equal(f(-42), "-42");
   test.equal(f(-4200000), "-4,200,000");
   test.equal(f(-42000000), "-42,000,000");
-  test.equal(f(1e21), "1e+21");
+  test.equal(f(1e21), "1,000,000,000,000,000,000,000");
+  test.equal(f(1.3e27), "1,300,000,000,000,000,044,761,612,288");
+  test.equal(f(1.3e107), "130,000,000,000,000,005,723,571,947,544,438,292,554,521,954,144,678,769,275,685,764,875,122,985,731,145,995,914,860,441,978,030,839,927,144,448");
```
Comment (Fil) [inline test/format-type-d-test.js:82]:
```suggestion
  test.equal(f(1.3e27).substring(0, 22), "1,300,000,000,000,000,");
  test.equal(f(1.3e107).substring(0, 22), "130,000,000,000,000,00");
  test.assert(f(1.3e107).match(/^[\d,]*$/));
```

or maybe
```
  test.assert(f(1.3e107).match(/^130,000,000,000,000,00[\d,]*$/));
```

Comment (mbostock) [reply test/format-type-d-test.js:82]: What’s wrong with the exact match?

**Comment by Fil** [comment]
In my understanding this issue is more about not switching to exponential notation for larger magnitudes than about accurately representing a number in a precision that doesn't exist.

This is for human consumption, and humans might say "Jeff B. has made 1,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000 dollars while you were brushing your teeth this morning"; but if that were an animated counter,  you'd prefer a random sequence of digits to create some movement.

In any case, I don't have a strong preference on how to represent the unknown part of the number (and if I had I would switch to either text or BigInt). This is why I suggest testing only the general format rather than the trailing (garbage) digits.

**Comment by mbostock** [comment]
I’d prefer to test the exact value in the sense that it’s a snapshot of the code’s actual behavior. Even if that behavior is not “correct”, it’ll be more obvious if we change the behavior as we’ll need to update the tests accordingly. I don’t feel updating the test here will be onerous or difficult to review, so I’d favor the explicit test.

That said, I just looked at #24 again and remembered that *number*.toLocaleString handles large numbers, too — so that’s probably what we should be using here instead of *number*.toPrecision or BigInt? Interestingly *number*.toLocaleString returns zeroes for the garbage digits, presumably because they figure that’s more likely what you intend (decimal instead of binary).

```js
(1.3e107).toLocaleString("en") 
```

So I’m thinking we use *number*.toLocaleString in the English locale and then strip the commas.

**Comment by mbostock** [comment]
Superseded by #104.

**Pull request closed without merging**